clang/lib/CodeGen/CGBuiltin.cpp

```diff
@@ -1060,238 +1061,331 @@ CodeGenFunction::evaluateOrEmitBuiltinObjectSize(const Expr *E, unsigned Type,
+
+  auto CheckForNegative = [&](Value *Res) {
+    Value *Cmp = Builder.CreateIsNotNeg(Res);
+    if (Index)
+      Cmp = Builder.CreateAnd(Builder.CreateIsNotNeg(Index), Cmp);
+    return Builder.CreateSelect(Cmp, Res,
+                                ConstantInt::get(ResType, 0, IsSigned));
+  };
 
-  // Build a load of the counted_by field.
-  bool IsSigned = CountedByFD->getType()->isSignedIntegerType();
-  Value *CountedByInst = EmitLoadOfCountedByField(Base, FAMDecl, CountedByFD);
-  if (!CountedByInst)
-    return getDefaultBuiltinObjectSizeResult(Type, ResType);
+  if (FlexibleArrayMemberFD == FD) {
+    if (Idx) { // Option (2) '&ptr->array[idx]'
+      //  size_t index_size = index * flexible_array_member_base_size;
+      Value *IndexSize = Builder.CreateMul(FlexibleArrayMemberBaseSize, Index,
+                                           "index_size", !IsSigned, IsSigned);
 
-  CountedByInst = Builder.CreateIntCast(CountedByInst, ResType, IsSigned);
+      //  return flexible_array_member_size - index_size;
+      return CheckForNegative(Builder.CreateSub(
+          FlexibleArrayMemberSize, IndexSize, "result", !IsSigned, IsSigned));
+    }
+  } else {
+    if (Idx) { // Option (4) '&ptr->field_array[idx]'
+      //  size_t field_base_size = sizeof (*ptr->field_array);
+      const ArrayType *ArrayTy = Ctx.getAsArrayType(FieldTy);
+      CharUnits BaseSize = Ctx.getTypeSizeInChars(ArrayTy->getElementType());
+      auto *FieldBaseSize =
+          llvm::ConstantInt::get(ResType, BaseSize.getQuantity(), IsSigned);
+
+      //  field_offset += index * field_base_size;
+      Value *Mul = Builder.CreateMul(Index, FieldBaseSize, "field_offset",
+                                     !IsSigned, IsSigned);
+      FieldOffset = Builder.CreateAdd(FieldOffset, Mul);
+
+      //  size_t offset_diff = flexible_array_member_offset - field_offset;
+      Value *OffsetDiff = Builder.CreateSub(SizeofStruct, FieldOffset,
+                                            "offset_diff", !IsSigned, IsSigned);
+
+      //  return offset_diff + flexible_array_member_size;
+      return CheckForNegative(
+          Builder.CreateAdd(FlexibleArrayMemberSize, OffsetDiff, "result"));
+    } else { // Option (3) '&ptr->field'
+      //  size_t offset_diff = flexible_array_member_offset - field_offset;
+      Value *OffsetDiff = Builder.CreateSub(SizeofStruct, FieldOffset,
+                                            "offset_diff", !IsSigned, IsSigned);
+
+      //  return flexible_array_member_size + offset_diff;
+      return CheckForNegative(
+          Builder.CreateAdd(FlexibleArrayMemberSize, OffsetDiff, "result"));
+    }
+  }
 
-  // Build a load of the index and subtract it from the count.
-  Value *IdxInst = nullptr;
-  if (Idx) {
-    if (Idx->HasSideEffects(getContext()))
-      // We can't have side-effects.
-      return getDefaultBuiltinObjectSizeResult(Type, ResType);
+  // Option (1) 'ptr->array'
```

Comment:
Nit: maybe you'd want to move this to `else` statement after `Option (2)`, for readability. Then, this fall through condition shouldn't be reachable.

Reply:
I'm fine with that. Should I put an "llvm_unreachable" at the end of the function?

Reply:
Yes please!

Reply:
Your wish is my command. :-)